python/pyiceberg/typedef.py

```diff
@@ -126,30 +135,41 @@ def json(self, exclude_none: bool = True, exclude: Optional[Set[str]] = None, by
 
 
 class PydanticStruct(IcebergBaseModel):
+    _position_to_field_name: Dict[int, str] = PrivateAttr()
+    _field_name_to_pydantic_field: Dict[str, Field] = PrivateAttr()
+
     class Config:
         frozen = False
 
-    def __setitem__(self, pos: int, value: Any) -> None:
-        positions = list(self.__fields__.values())
-        field = positions[pos]
-        if value is None:
-            if field.default is not None:
-                value = field.default
-            elif field.default_factory is not None:
-                value = field.default_factory()
+    @staticmethod
+    def _get_default_field_value(field: Field) -> Optional[Any]:
+        if field.default is not None:
+            return field.default
+        elif field.default_factory is not None:
+            return field.default_factory()
+        else:
+            return None
+
+    def set_record_schema(self, record_schema: StructType) -> None:
+        self._field_name_to_pydantic_field = {field.name: field for field in self.__fields__.values()}
+        self._position_to_field_name = {idx: field.name for idx, field in enumerate(record_schema.fields)}
+        for name, field in self.__fields__.items():
+            setattr(self, name, PydanticStruct._get_default_field_value(field))
 
-        self.__setattr__(field.name, value)
+    def __setitem__(self, pos: int, value: Any) -> None:
+        field_name = self._position_to_field_name[pos]
```

Comment:
Since it isn't safe to set defaults in this method, I think it can be simpler:

```python
    def __setitem__(self, pos: int, value: Any) -> None:
        self.__setattr__(self._position_to_field_name[pos], value)
```

Reply:
Indeed 👍🏻